### ISSUE TITLE
--no-snapshot --keep-source=0 does not delete all snapshots

### ISSUE THREAD

**Issue by fhriley** [psy0rz/zfs_autobackup#53]
Those options tell `zfs-autobackup` to not create a snapshot and keep 0 snapshots, but `zfs-autobackup` is [hardcoded](https://github.com/psy0rz/zfs_autobackup/blob/e65d1ac860cb8dcac0210fe09bf28cb9c235238c/bin/zfs-autobackup#L1113) to keep the last snapshot. `zfs-autobackup` should not override what I asked it to do.

**Comment by psy0rz** [comment]
True, this was originally to prevent people "shooting them self in the foot" and accidentally deleting the common snapshot.

However in snapshot-mode this makes sense i guess. And even if you want to do a one-time synchronization and delete the source snapshot after its complete?

Will fix this.

A --keep-target=0 doesn't make sense, right? :)

**Comment by psy0rz** [comment]
hmm.. i realise that keep-target=0 also is usefull for onetime replication .

**Comment by psy0rz** [comment]
I've made some preparations for this.

It will be called --keep-source=cleanup and --keep-target=cleanup, to indicate we really want a cleanup. E.g.: delete all snapshots belonging to this backup and breaking the backup.

Specifying 0 will become an error and inform the user of the cleanup if he really wants that.

**Comment by dreadedhamish** [comment]
> I've made some preparations for this.
> 
> It will be called --keep-source=cleanup and --keep-target=cleanup, to indicate we really want a cleanup. E.g.: delete all snapshots belonging to this backup and breaking the backup.
> 
> Specifying 0 will become an error and inform the user of the cleanup if he really wants that.

How will this "break the backup"?
I've got 1 drive that is space constrained, so I'd like to keep zero snapshots on the drive itself.

**Comment by Scrin** [comment]
With no snapshots zfs won't be able to know what has changed since the previous backup run, so the next backup would fail because the setup is effectively broken at that point, and fixing it usually requires destroying the target which is usually not desired

**Comment by dreadedhamish** [comment]
> With no snapshots zfs won't be able to know what has changed since the previous backup run, so the next backup would fail because the setup is effectively broken at that point, and fixing it usually requires destroying the target which is usually not desired

OK I think I get it - the primary snapshot can only be in the original pool - I can put a snapshot that automatically grows somewhere else. So my options are to increase the size of the original pool, or take more frequent snapshots so the primary growing one doesn't get too large.

**Comment by Scrin** [comment]
Yes, taking backups more frequently will allow the snapshot(s) on the source to stay smaller (also depends a lot on how the dataset is used, written to). The thinner can then take care of deleting unnecessary intermediate snapshots on the target so more frequent backup runs wont hurt your disk consumption there either

**Comment by psy0rz** [comment]
> How will this "break the backup"? I've got 1 drive that is space constrained, so I'd like to keep zero snapshots on the drive itself.

For an explanation see this: https://github.com/psy0rz/zfs_autobackup/wiki/Common-snapshots-and-holds

Also: bookmark support #23 would solve this, since they dont use space on the source.

**Comment by psy0rz** [comment]
In version v4.0 you can use --keep-source=0, it will use bookmarks. (which do not use space)
